### PR TITLE
[SPARK-30808][SQL] Enable Java 8 time API in Thrift server

### PR DESCRIPTION
**Pull request by MaxGekk**
### What changes were proposed in this pull request?
- Set `spark.sql.datetime.java8API.enabled` to `true` in `hiveResultString()`, and restore it back at the end of the call.
- Convert collected `java.time.Instant` & `java.time.LocalDate` to `java.sql.Timestamp` and `java.sql.Date` for correct formatting.

### Why are the changes needed?
Because of textual representation of timestamps/dates before 1582 year is incorrect:
```shell
$ export TZ="America/Los_Angeles"
$ ./bin/spark-sql -S
```
```sql
spark-sql> set spark.sql.session.timeZone=America/Los_Angeles;
spark.sql.session.timeZone	America/Los_Angeles
spark-sql> SELECT DATE_TRUNC('MILLENNIUM', DATE '1970-03-20');
1001-01-01 00:07:02
```
It must be 1001-01-01 00:**00:00**.

### Does this PR introduce any user-facing change?
Yes. After the changes:
```shell
$ export TZ="America/Los_Angeles"
$ ./bin/spark-sql -S
```
```sql
spark-sql> set spark.sql.session.timeZone=America/Los_Angeles;
spark.sql.session.timeZone	America/Los_Angeles
spark-sql> SELECT DATE_TRUNC('MILLENNIUM', DATE '1970-03-20');
1001-01-01 00:00:00
```

### How was this patch tested?
By running hive-thiftserver tests. In particular:
```
./build/sbt -Phadoop-2.7 -Phive-2.3 -Phive-thriftserver "hive-thriftserver/test:testOnly *SparkThriftServerProtocolVersionsSuite"
```

### CODE REVIEW

**Comment by MaxGekk** [comment]
@cloud-fan I tried to set the SQL config before the collect action in https://github.com/apache/spark/blob/aa0d13683cdf9f38f04cc0e73dc8cf63eed29bf4/sql/core/src/main/scala/org/apache/spark/sql/execution/HiveResult.scala#L54-L59 but it doesn't work - the SQL config didn't change the behavior :-(

File: sql/hive-thriftserver/src/main/scala/org/apache/spark/sql/hive/thriftserver/SparkExecuteStatementOperation.scala [hunk 1]
```diff
@@ -178,7 +179,12 @@ private[hive] class SparkExecuteStatementOperation(
           }
           curCol += 1
         }
-        resultRowSet.addRow(row.toArray.asInstanceOf[Array[Object]])
+        val resultRow = row.map {
```
Comment (MaxGekk) [inline sql/hive-thriftserver/src/main/scala/org/apache/spark/sql/hive/thriftserver/SparkExecuteStatementOperation.scala]: I have to convert Java 8 classes to `java.sql.Timestamp`/`Date` for correct formatting. I cannot use `TimestampFormatter` here because the conversions to strings can be performed in `ColumnBuffer` in `hive-serde`.

**Comment by cloud-fan** [comment]
> but it doesn't work - the SQL config didn't change the behavior

It's probably because the `Dataset` is already there and its encoder is already created, which decides to return old timestamp/date. We can create a new `Dataset` to fix it, e.g. `df.filter(lit("true")).collect`

**Comment by MaxGekk** [comment]
> We can create a new Dataset to fix it ...

@cloud-fan Is it possible to create/restore a Dataset from an executedPlan?

**Comment by cloud-fan** [comment]
@MaxGekk it's possible from a logical plan, e.g. `Dataset.ofRows(session, df.logicalPlan)`

**Comment by MaxGekk** [comment]
jenkins, retest this, please

File: sql/core/src/main/scala/org/apache/spark/sql/execution/HiveResult.scala [hunk 2]
```diff
@@ -36,27 +37,40 @@ object HiveResult {
...
+              Option(comment.asInstanceOf[String]).getOrElse(""))
+              .map(s => String.format(s"%-20s", s))
+              .mkString("\t")
+        }
+      // SHOW TABLES in Hive only output table names,
+      // while ours output database, table name, isTemp.
+      case command @ ExecutedCommandExec(s: ShowTablesCommand) if !s.isExtended =>
+        command.executeCollect().map(_.getString(1))
+      case _ =>
+        val java8TimeConf = SQLConf.get.getConf(SQLConf.DATETIME_JAVA8API_ENABLED)
```
Comment (cloud-fan) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/HiveResult.scala]: shall we follow `QueryExecution.preparations`?
```
val sessionWithJava8DatatimeEnabled = {
  val cloned = ds.session.cloneSession()
  cloned.... setConf
  cloned
}
cloned.withActive {
  // the actual work
}
```

File: sql/core/src/main/scala/org/apache/spark/sql/execution/HiveResult.scala [hunk 3]
```diff
@@ -36,27 +37,40 @@ object HiveResult {
...
+        try {
+          SQLConf.get.setConf(SQLConf.DATETIME_JAVA8API_ENABLED, true)
+          val result: Seq[Seq[Any]] = ds.filter(lit(true))
```
Comment (cloud-fan) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/HiveResult.scala]: how about `Dataset.ofRows(session, ds.queryExecution.logicalPlan)`?

File: sql/core/src/main/scala/org/apache/spark/sql/execution/HiveResult.scala [hunk 4]
```diff
@@ -36,27 +37,40 @@ object HiveResult {
...
-    case command @ ExecutedCommandExec(s: ShowTablesCommand) if !s.isExtended =>
-      command.executeCollect().map(_.getString(1))
-    case other =>
-      val result: Seq[Seq[Any]] = other.executeCollectPublic().map(_.toSeq).toSeq
-      // We need the types so we can output struct field names
-      val types = executedPlan.output.map(_.dataType)
-      // Reformat to match hive tab delimited output.
-      result.map(_.zip(types).map(e => toHiveString(e)))
-        .map(_.mkString("\t"))
+  def hiveResultString[T](ds: Dataset[T]): Seq[String] = {
```
Comment (cloud-fan) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/HiveResult.scala]: how about `def hiveResultString(ds: Dataset[_])`?

File: sql/hive-thriftserver/src/main/scala/org/apache/spark/sql/hive/thriftserver/SparkExecuteStatementOperation.scala (hunk 1 above)
Comment (cloud-fan) [reply sql/hive-thriftserver/src/main/scala/org/apache/spark/sql/hive/thriftserver/SparkExecuteStatementOperation.scala]: can you add a code comment? Say that the results are passed to Hive and Hive only accepts  `Timestamp`/`Date`

File: sql/core/src/main/scala/org/apache/spark/sql/execution/HiveResult.scala [hunk 5]
```diff
@@ -36,27 +37,41 @@ object HiveResult {
...
+      case command @ ExecutedCommandExec(s: ShowTablesCommand) if !s.isExtended =>
+        command.executeCollect().map(_.getString(1))
+      case _ =>
+        val sessionWithJava8DatetimeEnabled = {
+          val cloned = ds.sparkSession.cloneSession()
+          cloned.conf.set(SQLConf.DATETIME_JAVA8API_ENABLED.key, true)
+          cloned
+        }
+        sessionWithJava8DatetimeEnabled.withActive {
+          val result: Seq[Seq[Any]] = Dataset.ofRows(ds.sparkSession, ds.queryExecution.logical)
```
Comment (cloud-fan) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/HiveResult.scala]: nit: let's add a code comment to explain why we can't collect the input `ds`. Its encoder was already created with java 8 datetime disabled.

File: sql/hive-thriftserver/src/main/scala/org/apache/spark/sql/hive/thriftserver/SparkSQLDriver.scala [hunk 6]
```diff
@@ -60,9 +60,10 @@ private[hive] class SparkSQLDriver(val context: SQLContext = SparkSQLEnv.sqlCont
     // TODO unify the error code
     try {
       context.sparkContext.setJobDescription(command)
+      val df = context.sql(command)
       val execution = context.sessionState.executePlan(context.sql(command).logicalPlan)
       hiveResponse = SQLExecution.withNewExecutionId(execution) {
```
Comment (cloud-fan) [inline sql/hive-thriftserver/src/main/scala/org/apache/spark/sql/hive/thriftserver/SparkSQLDriver.scala]: `SQLExecution.withNewExecutionId(df.queryExecution)`?

File: sql/hive-thriftserver/src/main/scala/org/apache/spark/sql/hive/thriftserver/SparkSQLEnv.scala [hunk 7]
```diff
@@ -24,6 +24,7 @@ import org.apache.spark.{SparkConf, SparkContext}
 import org.apache.spark.internal.Logging
 import org.apache.spark.sql.{SparkSession, SQLContext}
 import org.apache.spark.sql.hive.{HiveExternalCatalog, HiveUtils}
+import org.apache.spark.sql.internal.SQLConf
```
Comment (cloud-fan) [inline sql/hive-thriftserver/src/main/scala/org/apache/spark/sql/hive/thriftserver/SparkSQLEnv.scala]: unnecessary change?

File: sql/hive-thriftserver/src/main/scala/org/apache/spark/sql/hive/thriftserver/SparkSQLDriver.scala [hunk 8]
```diff
@@ -60,9 +60,10 @@ private[hive] class SparkSQLDriver(val context: SQLContext = SparkSQLEnv.sqlCont
     // TODO unify the error code
     try {
       context.sparkContext.setJobDescription(command)
+      val df = context.sql(command)
       val execution = context.sessionState.executePlan(context.sql(command).logicalPlan)
```
Comment (cloud-fan) [inline sql/hive-thriftserver/src/main/scala/org/apache/spark/sql/hive/thriftserver/SparkSQLDriver.scala]: `val execution = df.queryExecution`?

File: sql/hive/src/test/scala/org/apache/spark/sql/hive/test/TestHive.scala [hunk 9]
```diff
@@ -605,6 +605,8 @@ private[hive] class TestHiveQueryExecution(
     // Proceed with analysis.
     sparkSession.sessionState.analyzer.executeAndCheck(logical, tracker)
   }
+
+  lazy val dataset = Dataset.ofRows(sparkSession, logical)
```
Comment (cloud-fan) [inline sql/hive/src/test/scala/org/apache/spark/sql/hive/test/TestHive.scala]: If it's only used once in https://github.com/apache/spark/pull/27552/files#diff-6b519cd0a9dc5b0ee9caab7712d252e9R349 , let's just inline it.

**Review by cloud-fan — APPROVED** [review]
LGTM except a few minor comments

**Comment by MaxGekk** [comment]
So many HiveComparisonTest related tests failed, I will revert this https://github.com/apache/spark/pull/27552/commits/cdb322d6366e5c92579a8eaa26ccd94626cdf3ee

**Comment by MaxGekk** [comment]
@cloud-fan Something wrong is going on here. Commands issues from HiveComparisonTest are executed twice, it seems.

**Comment by cloud-fan** [comment]
@MaxGekk yea creating the df again may execute the command again. Let's keep the lazy val.

**Comment by MaxGekk** [comment]
I would prefer to cut off this PR at this point https://github.com/apache/spark/pull/27552#issuecomment-585664688, and implement moving settings of `spark.sql.datetime.java8API.enabled` to hiveResult() separately.

Making dataset as lazy val doesn't help me, so, I stuck for now.

**Comment by MaxGekk** [comment]
While debugging the `test ambiguousReferences resolved as hive` test from `org.apache.spark.sql.hive.execution.HiveResolutionSuite`, I see CreateDataSourceTableCommand is invoke twice for the same table name **t1** :
1. In `lazy val dataset = Dataset.ofRows(sparkSession, logical)` inside of `TestHiveQueryExecution`
2. In `hiveResultString()`:
```scala
val result: Seq[Seq[Any]] = Dataset.ofRows(ds.sparkSession, ds.queryExecution.logical)
            .queryExecution
            .executedPlan
            .executeCollectPublic().map(_.toSeq).toSeq
```

**Comment by cloud-fan** [comment]
@MaxGekk I see the problem now. We should use `ds.logicalPlan` instead of `ds.queryExecution.logical`. As `ds.logicalPlan` is `LocalRelation` to hold the command result.

**Comment by cloud-fan** [comment]
thanks, merging to master/3.0!

File: sql/core/src/main/scala/org/apache/spark/sql/execution/HiveResult.scala [hunk 10]
```diff
@@ -36,27 +37,43 @@ object HiveResult {
...
+      case _ =>
+        val sessionWithJava8DatetimeEnabled = {
+          val cloned = ds.sparkSession.cloneSession()
+          cloned.conf.set(SQLConf.DATETIME_JAVA8API_ENABLED.key, true)
+          cloned
+        }
+        sessionWithJava8DatetimeEnabled.withActive {
+          // We cannot collect the original dataset because its encoders could be created
+          // with disabled Java 8 date-time API.
+          val result: Seq[Seq[Any]] = Dataset.ofRows(ds.sparkSession, ds.logicalPlan)
```
Comment (cloud-fan) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/HiveResult.scala]: found a problem. `Dataset.ofRows` will set the input session as active, so we should write `Dataset.ofRows(sessionWithJava8DatetimeEnabled, ...` and remove the outer `sessionWithJava8DatetimeEnabled.withActive`.

File: sql/core/src/test/scala/org/apache/spark/sql/SQLQueryTestSuite.scala [hunk 11]
```diff
@@ -512,7 +511,7 @@ class SQLQueryTestSuite extends QueryTest with SharedSparkSession {
     val schema = df.schema.catalogString
     // Get answer, but also get rid of the #1234 expression ids that show up in explain plans
     val answer = SQLExecution.withNewExecutionId(df.queryExecution, Some(sql)) {
-      hiveResultString(df.queryExecution.executedPlan).map(replaceNotIncludedMsg)
+      hiveResultString(df).map(replaceNotIncludedMsg)
```
Comment (cloud-fan) [inline sql/core/src/test/scala/org/apache/spark/sql/SQLQueryTestSuite.scala]: in `ThriftServerQueryTestSuite`, we get the result by JDBC, so there is no DataFrame created.

We should follow pgsql and return java 8 datetime when the config is enabled. https://jdbc.postgresql.org/documentation/head/8-date-time.html

cc @wangyum @yaooqinn

Comment (yaooqinn) [reply sql/core/src/test/scala/org/apache/spark/sql/SQLQueryTestSuite.scala]: We do not have our own JDBC implementation yet, and the hive JDBC haven't supported this too
https://github.com/apache/hive/blob/ebd41526b9433a196ae0f934bc466cb5c155a202/jdbc/src/java/org/apache/hive/jdbc/HiveQueryResultSet.java#L402-L411

File: sql/hive-thriftserver/src/main/scala/org/apache/spark/sql/hive/thriftserver/SparkExecuteStatementOperation.scala [hunk 12]
```diff
@@ -178,7 +179,14 @@ private[hive] class SparkExecuteStatementOperation(
           }
           curCol += 1
         }
-        resultRowSet.addRow(row.toArray.asInstanceOf[Array[Object]])
+        // Convert date-time instances to types that are acceptable by Hive libs
+        // used in conversions to strings.
+        val resultRow = row.map {
+          case i: Instant => Timestamp.from(i)
```
Comment (yaooqinn) [inline sql/hive-thriftserver/src/main/scala/org/apache/spark/sql/hive/thriftserver/SparkExecuteStatementOperation.scala]: There seems no java8 values to be add to the `row` buffer here by `SparkExecuteStatementOperation#addNonNullColumnValue `

https://github.com/apache/spark/pull/27552/files#diff-72dcd8f81a51c8a815159fdf0332acdcR84-R116

Comment (cloud-fan) [reply sql/hive-thriftserver/src/main/scala/org/apache/spark/sql/hive/thriftserver/SparkExecuteStatementOperation.scala]: can you help fix it? I think we should output java8 datetime values if the config is enabled.

Comment (yaooqinn) [reply sql/hive-thriftserver/src/main/scala/org/apache/spark/sql/hive/thriftserver/SparkExecuteStatementOperation.scala]: We are limited by `hive-jdbc` module, see https://github.com/apache/hive/blob/a7e704c679a00db68db9b9f921d133d79a32cfcc/jdbc/src/java/org/apache/hive/jdbc/HiveBaseResultSet.java#L427-L457, we might need our own jdbc driver implementation to achieve this

File: sql/core/src/main/scala/org/apache/spark/sql/execution/HiveResult.scala [hunk 13]
```diff
@@ -36,27 +37,43 @@ object HiveResult {
...
+              .mkString("\t")
+        }
+      // SHOW TABLES in Hive only output table names,
+      // while ours output database, table name, isTemp.
+      case command @ ExecutedCommandExec(s: ShowTablesCommand) if !s.isExtended =>
+        command.executeCollect().map(_.getString(1))
+      case _ =>
+        val sessionWithJava8DatetimeEnabled = {
+          val cloned = ds.sparkSession.cloneSession()
+          cloned.conf.set(SQLConf.DATETIME_JAVA8API_ENABLED.key, true)
```
Comment (yaooqinn) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/HiveResult.scala]: why is this always `true`?

Comment (cloud-fan) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/HiveResult.scala]: the old `Date`/`Timestamp` doesn't follow the new calendar and may produce wrong string for some date/timestamp values.

Comment (cloud-fan) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/HiveResult.scala]: oh wait, we format `Date/Timestamp` by our own formatter, so this should be no problem.

Comment (yaooqinn) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/HiveResult.scala]:
```
-- !query
set spark.sql.datetime.java8API.enabled
-- !query schema
struct<key:string,value:string>
-- !query output
spark.sql.datetime.java8API.enabled	false


-- !query
set set spark.sql.session.timeZone=America/Los_Angeles
-- !query schema
struct<key:string,value:string>
-- !query output
set spark.sql.session.timeZone	America/Los_Angeles


-- !query
SELECT DATE_TRUNC('MILLENNIUM', DATE '1970-03-20')
-- !query schema
struct<date_trunc(MILLENNIUM, CAST(DATE '1970-03-20' AS TIMESTAMP)):timestamp>
-- !query output
1001-01-01 00:00:00
```

I rm this line and run `SQLQueryTestSuite` with cases above,  the results are the same. Or does this problem only exists for  `spark-sql` script?

Comment (MaxGekk) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/HiveResult.scala]: > Or does this problem only exists for spark-sql script?

Only when thrift-server is involved in the loop.

Comment (yaooqinn) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/HiveResult.scala]: I also pass these test through `ThriftServerQueryTestSuite`

Comment (yaooqinn) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/HiveResult.scala]:
```
spark-sql> set spark.sql.session.timeZone=America/Los_Angeles;
spark.sql.session.timeZone	America/Los_Angeles
spark-sql> SELECT DATE_TRUNC('MILLENNIUM', DATE '1970-03-20');
1001-01-01 00:00:00
spark-sql> select version();
3.1.0 b3dcb63a682bc31827a86cf381f157a81e9e07ac
```
Also correct with `bin/spark-sql`

Comment (cloud-fan) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/HiveResult.scala]: Yea I tested it too and looks fine. Maybe some refactor of how to format old `Date/Timestamp` fixes it already.

@yaooqinn can you send a PR to revert it? Let's see if all tests pass.

Comment (yaooqinn) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/HiveResult.scala]: OK
